test/Microsoft.ML.Tests/BinaryLoaderSaverTests.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using Microsoft.ML.Data;
+using Microsoft.ML.Data.IO;
+using Microsoft.ML.RunTests;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.ML.Tests
+{
+    public class BinaryLoaderSaverTests : TestDataPipeBase
+    {
+        public BinaryLoaderSaverTests(ITestOutputHelper output) : base(output)
+        {
+        }
+
+        [Fact]
+        public void OldKeyTypeCodecTest()
+        {
+            // Checks that we can load IDataViews defined with unknown cardinality KeyType.
+            // schema-codec-test.idv was generated with the following command before simplifying the KeyType:
+            // dotnet MML.dll savedata loader=text{col=A:U4[0-2]:0 col=B:U4[2-5]:0 col=C:U1[5-10]:0 col=D:U2[0-*]:0 col=E:U4[4-*]:0 col=F:U8[5-*]:0} dout=codectest.idv
```

Comment:
>col=A:U4[0-2]:0 col=B:U4[2-5]:0 col=C:U1[5-10]:0 col=D:U2[0-*]:0 col=E:U4[4-*]:0 col=F:U8[5-* [](start = 51, length = 93)

So right now this correspond to this.

```
col=A:U4[2]:0
col=B:U4[3]:1
col=C:U1[5]:2
col=D:U2[65534]:3
col=E:U4[4294967294]:4
col=F:U8[18446744073709551614]:5
```

Now here I am wondering something. We had previously that if a specified a range, the range was inclusive -- that is, `0-2` being 0,1,2, has a count of 3. I wonder if it would be more clear if in the "single number" case, whether this indicates a count or a range. Because if I see `U4[2]` I might think this is a key value with 2 values, but in fact, there are 3, because that 2 is merely an upper bound. This at least strikes me as confusing. So we might read `U4[0-2]` but it is equivalent to `U4[3]`. Not sure. Maybe ask opinion of @yaeldekel and @wschin.

Reply:
Actually yeah the more I think about it the more confused I get... so in the new scheme, `U4[2]` would correspond to a count of *three* where the actual physical values corresponding to valid values are `1`, `2`, and `3`. So I think in this "single value" case it should just be count.

---
In reply to: [249994613](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249994613) [](ancestors = 249994613)

Reply:
As you wrote in your message, there is a fundamental difference: are we using the `Count` as a max (for a range) or as a count of values. In this change, I use Count as a the upper bound of the range of valid values, and the upper bound is inclusive.

In fact, Wei-Sheng suggested to rename Count to Max.

---
In reply to: [249996684](https://github.com/dotnet/machinelearning/pull/2146#discussion_r249996684) [](ancestors = 249996684,249994613)

Reply:
I have initially worked on removing `KeyRange`. I succeeded in removing it completely, but I found one scenario that is not allowed without `KeyRange`. 

Both for TextLoader.Column and for TypeConvertingTransformer.ColumnInfo we used to define a `KeyRange`. We used to check if the Column defines a key by checking whether the `KeyRange `is null. If it is defined but no `Max` is provided we would map to `kind.ToMaxInt()` when constructing a `KeyType `from the `KeyRange`.

In my change I had replaced `KeyRange` with `ulong? KeyCount`. This simplified nicely some of the code since KeyRange used to define a `Max` which was essentially acting as `KeyCount - 1`. However, we do not distinguish the case `KeyRange == null` and `KeyRange != null && KeyRange.Max == null`. In other words `ulong? KeyCount == null` could mean that the current column does not define a key, or that the current column defines a key, but that its `keyCount `is not defined and should be mapped to `kind.ToMaxInt()`.

I therefore decided for a compromise. I transformed `KeyRange` into `KeyCount`, and implemented the logic that makes the `Count` of `KeyCount` match `KeyType.Count`.

Reply:
Could we use `Count=0` to denote `kind.ToMaxInt()`?

---
In reply to: [250095679](https://github.com/dotnet/machinelearning/pull/2146#discussion_r250095679) [](ancestors = 250095679)